### DISCUSSION TITLE
Caching and pending Image Transform URLs

### DISCUSSION THREAD

**Discussion by Michael-Paragonn** [craftcms/cms#16195] [Ideas] [answered]
Most sites I work on use Cloudflare for caching. However, CF has a nasty habit of caching _pending image transform URLs_ (`example.com/actions/assets/generate-transform?transformId=123`). To work around that, we turn on `generateTransformsBeforePageLoad` in the config, but that has the unfortunate side effect of massively delaying page rendering under certain situations (devs loading up a clone of a production site locally, with image transforms stored in a remote CDN, is one of the situations where page loading can take upwards of 5 minutes at times while the transforms are generated).

Anyway, it seems like there should be a more direct and simple solution. I'm not the most knowledgeable about them, but I know web requests can return page headers. One of these is `Cache-Control: no-store, no-cache, must-revalidate`... or if the page _should_ be cached, it has "cache-for-this-long" headers.

I also know that Craft's native `{% cache %}` tag detects pending image transform URLs, and doesn't cache a page if they're present.

Presumably, Cloudflare honors a page's `no-cache` header. So why doesn't Craft force a `Cache-Control: no-store, no-cache, must-revalidate` header if there are any pending image transform URLs present in the page?

**Comment by brandonkelly** [comment, marked as answer]
Good suggestion! Just made this change for Craft 4.14 and 5.6. (d7dce98aecc29480d2f629624b0fe981c3929e0c)

**Reply by brandonkelly** [reply]
Craft 4.14.0 and 5.6.0 are out with that change. Thanks again!

**Reply by Michael-Paragonn** [reply]
Thank _you_ - this is wonderful!
